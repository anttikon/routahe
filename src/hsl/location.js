import fetch from 'node-fetch'
import querystring from 'querystring'

const getHslLocation = async (query) => {
  const params = {
    text: query,
    'boundary.rect.min_lat': 59.9,
    'boundary.rect.max_lat': 60.45,
    'boundary.rect.min_lon': 24.3,
    'boundary.rect.max_lon': 25.5,
    size: 1,
    lang: 'fi',
  }
  const response = await fetch(`http://api.digitransit.fi/geocoding/v1/search?${querystring.stringify(params)}`)
  return response.json()
}

const mapFeature = feature => {
  if (!feature) {
    return feature
  }
  const [lon, lat] = feature.geometry.coordinates
  const label = feature.properties.label
  return { label, lon, lat }
}

export const getLocationByString = async query => {
  const hslLocation = await getHslLocation(query)
  return mapFeature(hslLocation.features?.[0])
}
