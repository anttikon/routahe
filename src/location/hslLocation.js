import fetch from 'node-fetch'
import {head} from 'lodash'

async function getHslLocations(query) {
  const response = await fetch(getUrl('http://api.digitransit.fi/geocoding/v1/search', {
    'text': encodeURIComponent(query),
    'boundary.rect.min_lat': 59.9,
    'boundary.rect.max_lat': 60.45,
    'boundary.rect.min_lon': 24.3,
    'boundary.rect.max_lon': 25.5
  }))
  return response.json()
}

async function getLocationByQuery(query) {
  const hslLocations = await getHslLocations(query)
  if (!hslLocations.features) {
    return {}
  }
  return head(hslLocations.features.map(mapFeature))

  function mapFeature(feature) {
    const [lon, lat] = feature.geometry.coordinates
    const label = feature.properties.label
    return {label, lon, lat}
  }
}

function getUrl(api, params) {
  const urlParams = Object.keys(params).map(key => `${key}=${params[key]}`).join('&')
  return `${api}?${urlParams}`
}


export {getHslLocations, getLocationByQuery}