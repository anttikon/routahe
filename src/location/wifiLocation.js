import triangulate from 'wifi-triangulate'
import fetch from 'node-fetch'
import {head} from 'lodash'

exports.getLocationByWifi = async() => {
  const {lat, lng, accuracy} = await getMyLocation()
  const label = await exports.getStreetAddressByLocation(lat, lng)
  return {lon: lng, lat, label, accuracy,}

  function getMyLocation() {
    return new Promise((resolve, reject) => triangulate((err, location) => err ? reject(err) : resolve(location)))
  }
}

exports.getStreetAddressByLocation = async(lat, lon) => {
  const response = await fetch(`http://maps.googleapis.com/maps/api/geocode/json?latlng=${lat},${lon}&sensor=true`)
  const {formatted_address} = head((await response.json()).results)
  return formatted_address || 'Unknown'
}