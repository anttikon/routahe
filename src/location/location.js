import {getLocationByQuery} from './hslLocation'
import {getLocationByWifi} from './wifiLocation'

exports.getLocation = (searchTerm) => searchTerm ? getLocationByQuery(searchTerm) : getLocationByWifi()