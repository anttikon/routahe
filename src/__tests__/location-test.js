import { mapFeature } from '../location'

const dummyFeature = {
  type: 'Feature',
  geometry: { type: 'Point', coordinates: [24.943871, 60.158048] },
  properties: {
    id: 'node:4950866718',
    gid: 'openstreetmap:address:node:4950866718',
    layer: 'address',
    source: 'openstreetmap',
    source_id: 'node:4950866718',
    name: 'Tehtaankatu 24',
    housenumber: '24',
    street: 'Tehtaankatu',
    postalcode: '00150',
    postalcode_gid: 'whosonfirst:postalcode:421472993',
    confidence: 0.7516263013376072,
    accuracy: 'point',
    country: 'Suomi',
    country_gid: 'whosonfirst:country:0',
    country_a: 'FIN',
    region: 'Uusimaa',
    region_gid: 'whosonfirst:region:85683067',
    localadmin: 'Helsinki',
    localadmin_gid: 'whosonfirst:localadmin:907199715',
    locality: 'Helsinki',
    locality_gid: 'whosonfirst:locality:101748417',
    neighbourhood: 'Ullanlinna',
    neighbourhood_gid: 'whosonfirst:neighbourhood:85907973',
    label: 'Tehtaankatu 24, Helsinki',
  },
}

describe('mapFeature', () => {
  it('should return {label, lon, lat}', () => {
    expect(mapFeature(dummyFeature)).toMatchSnapshot()
  })
})
