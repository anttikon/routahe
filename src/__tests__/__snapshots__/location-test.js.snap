// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapFeature should return {label, lon, lat} 1`] = `
Object {
  "label": "Tehtaankatu 24, Helsinki",
  "lat": 60.158048,
  "lon": 24.943871,
}
`;
